examples/snake/snake.cc

```diff
@@ -0,0 +1,651 @@
+// Copyright lowRISC Contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cheri.hh>
+#include <compartment.h>
+#include <debug.hh>
+#include <platform-gpio.hh>
+#include <thread.h>
+#include <vector>
+
+#include "../../libraries/lcd.hh"
+#include "cherry_bitmap.h"
+#include "xor_shift.hh"
+
+using Debug = ConditionalDebug<true, "Snake">;
+using namespace sonata::lcd;
+using namespace CHERI;
+
+// Control game speed
+static constexpr uint32_t MILLISECONDS_PER_FRAME = 400;
+// Small wait between games to avoid accidentally starting the next
+static constexpr uint32_t START_MENU_WAIT_MILLISECONDS = 400;
+
+// Snake speeds up as it gets longer if enabled
+static constexpr bool SPEED_SCALING_ENABLED = true;
+// If enabled, all joystick motions start the game (not just a press)
+static constexpr bool START_ON_ANY_INPUT = true;
+// If enabled, displays a cherry bitmap for the fruit instead of a
+// green square at 10x10 and 5x5 tile sizes.
+static constexpr bool USE_CHERRY_IMAGE = true;
+
+// Change colour of game elements
+static constexpr Color BACKGROUND_COLOR = Color::Black,
+                       BORDER_COLOR     = Color::White,
+                       FOREGROUND_COLOR = Color::White,
+                       SNAKE_COLOR      = Color::Red;
+// Change the size of game elements (will automatically fit display)
+static constexpr Size TILE_SIZE    = {10, 10}, 
+                      TILE_SPACING = { 2,  2},
+                      BORDER_SIZE  = { 4,  3};
+
+typedef struct Position
+{
+	int32_t x;
+	int32_t y;
+} Position;
+
+enum class Direction
+{
+	UP    = 0,
+	RIGHT = 1,
+	DOWN  = 2,
+	LEFT  = 3
+};
+
+enum class Tile : uint64_t
+{
+	EMPTY,
+	SNAKE,
+	FRUIT
+};
+
+/**
+ * @brief Converts a given size_t to an equivalent string representing its
+ * unsigned base 10 representation in the given buffer.
+ *
+ * @param buffer The buffer/string to write the converted number to. Will
+ * terminate at the end of the string.
+ * @param num The size_t number to convert
+ */
+void size_t_to_str_base10(char *buffer, size_t num)
+{
+	// Parse the digits using repeated remainders mod 10
+	int8_t end_idx = 0;
```

Comment:
```suggestion
	size_t end_idx = 0;
```
Use `size_t` for array indices (or ptrdiff_t if the index can go negative)